File: android/app/src/main/java/com/wispralt/keyboard/AuthClient.kt
package com.wispralt.keyboard

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Thin OkHttp wrapper around /auth/request and /auth/verify on the
 * backend. Pure I/O, no Android types — call from a background thread.
 */
object AuthClient {
    private val http = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private val JSON = "application/json".toMediaType()

    sealed class Result<out T> {
        data class Ok<T>(val value: T) : Result<T>()
        data class Err(val message: String) : Result<Nothing>()
    }

    /** POST /auth/request — emails a 6-digit code to [email]. */
    fun requestCode(email: String): Result<Unit> {
        val body = JSONObject().put("email", email).toString()
            .toRequestBody(JSON)
        val req = Request.Builder()
            .url("${BuildConfig.BACKEND_URL}/auth/request")
            .post(body)
            .build()
        return runCatching {
            http.newCall(req).execute().use { res ->
                if (res.isSuccessful) Result.Ok(Unit)
                else Result.Err(parseError(res.body?.string()) ?: "HTTP ${res.code}")
            }
        }.getOrElse { Result.Err(it.message ?: "network error") }
    }

    data class Session(val token: String, val email: String)

    /** POST /auth/verify — exchanges the 6-digit code for a JWT. */
    fun verifyCode(email: String, code: String): Result<Session> {
        val body = JSONObject()
            .put("email", email)
            .put("code", code)
            .toString()
            .toRequestBody(JSON)
        val req = Request.Builder()
            .url("${BuildConfig.BACKEND_URL}/auth/verify")
            .post(body)
            .build()
        return runCatching {
            http.newCall(req).execute().use { res ->
                val text = res.body?.string()
                if (!res.isSuccessful) {
                    return Result.Err(parseError(text) ?: "HTTP ${res.code}")
                }
                val json = JSONObject(text ?: "{}")
                val token = json.optString("token").ifBlank { return Result.Err("missing token") }
                val user = json.optJSONObject("user")
                val verifiedEmail = user?.optString("email").orEmpty()
                    .ifBlank { email }
                Result.Ok(Session(token, verifiedEmail))
            }
        }.getOrElse { Result.Err(it.message ?: "network error") }
    }

    private fun parseError(body: String?): String? {
        if (body.isNullOrBlank()) return null
        return runCatching { JSONObject(body).optString("error").takeIf { it.isNotBlank() } }
            .getOrNull()
    }
}

File: ios/wispr-alt/AuthClient.swift
import Foundation

/// Thin URLSession wrapper around /auth/request and /auth/verify on the
/// backend. Pure async functions, no UI types — call from any context.
enum AuthClient {
    struct Session {
        let token: String
        let email: String
    }

    enum AuthError: LocalizedError {
        case server(String)
        case network(String)

        var errorDescription: String? {
            switch self {
            case .server(let m), .network(let m): return m
            }
        }
    }

    /// POST /auth/request — emails a 6-digit code to [email].
    static func requestCode(email: String) async throws {
        let url = URL(string: "\(Backend.url)/auth/request")!
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(
            withJSONObject: ["email": email],
            options: []
        )

        let (data, resp) = try await URLSession.shared.data(for: req)
        try ensureOk(data: data, resp: resp)
    }

    /// POST /auth/verify — exchanges the code for a JWT.
    static func verifyCode(email: String, code: String) async throws -> Session {
        let url = URL(string: "\(Backend.url)/auth/verify")!
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(
            withJSONObject: ["email": email, "code": code],
            options: []
        )

        let (data, resp) = try await URLSession.shared.data(for: req)
        try ensureOk(data: data, resp: resp)

        struct VerifyResponse: Decodable {
            let token: String
            let user: User
            struct User: Decodable { let id: String; let email: String }
        }
        let decoded = try JSONDecoder().decode(VerifyResponse.self, from: data)
        return Session(token: decoded.token, email: decoded.user.email)
    }

    private static func ensureOk(data: Data, resp: URLResponse) throws {
        guard let http = resp as? HTTPURLResponse else {
            throw AuthError.network("invalid response")
        }
        if (200..<300).contains(http.statusCode) { return }
        // Try to surface the backend's `error` field; fall back to status code.
        if let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let msg = obj["error"] as? String, !msg.isEmpty {
            throw AuthError.server(msg)
        }
        throw AuthError.server("HTTP \(http.statusCode)")
    }
}
